### PR TITLE
fix: Text cut off issues when adjusting text size and font weight in system settings

### PR DESCRIPTION
**Pull request by jcdhlzq**
Fix Text cut off issues when adjusting text size and font weight in system settings.

## Summary:

This pr fixed the problem that can be reproduced with the [snack](https://snack.expo.dev/fl5DSrLBJ) on Xiaomi devices with MIUI13 and MIUI14. The problem is shown as the image below: the number "999" is cut off and only "99" is rendered.
text-cutoff-when-scaling-miui14

The problem is produced with setting font scaling in system settings like the image below shows.

settings 

This text cut off case can be avoided by setting [`allowFontScaling`](https://reactnative.cn/docs/text#allowfontscaling) to false. But this pr can make it no matter what value `allowFontScaling` is set.

The root cause of this case, according to MIUI developers, is that Misans typeface, as a variable font which will adjust the weight of different font axes with different font size, gets different widths when Text is measured and drawn for a bug in the framework of MIUI rom. They will fix this bug in next version while this pr fixed it in old versions.

## Changelog:

[ANDROID][FIXED]-Fix Text cut off issues when adjusting text size and font weight in system settings.

## Test Plan:

The effect after fixed is shown in the image below.

fixed

### CODE REVIEW

**Comment by ryancat** [comment]
It would be helpful if we could have a test in RNTester for this. Does this only reproduce in Xiaomi devices?

**Comment by jcdhlzq** [comment]
> It would be helpful if we could have a test in RNTester for this. Does this only reproduce in Xiaomi devices?

Yes, only Xiaomi devices with MIUI13 and MIUI14 devices are reported from users in product environment until now.

As for the test, there is a snack demo in this pr's summary, [demo](https://snack.expo.dev/fl5DSrLBJ)

**Comment by jcdhlzq** [comment]
> It would be helpful if we could have a test in RNTester for this. Does this only reproduce in Xiaomi devices?

Some test cases are added in a seperate pr: #39594

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java [hunk 1]
```diff
@@ -680,16 +666,37 @@ protected boolean dispatchHoverEvent(MotionEvent event) {
     return super.dispatchHoverEvent(event);
   }
 
-  private void applyTextAttributes() {
-    // Workaround for an issue where text can be cut off with an ellipsis when
-    // using certain font sizes and padding. Sets the provided text size and
-    // letter spacing to ensure consistent rendering and prevent cut-off.
-    if (!Float.isNaN(mFontSize)) {
-      setTextSize(TypedValue.COMPLEX_UNIT_PX, mFontSize);
+  public void setLetterSpacingPt(float letterSpacingPt) {
```
Comment (NickGerleman) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java:669]: Why was this changed to `pt`? Input should be SP/dips.

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java [hunk 2]
```diff
@@ -67,6 +67,12 @@ public class ReactTextView extends AppCompatTextView implements ReactCompoundVie
   private ReactViewBackgroundManager mReactBackgroundManager;
   private Spannable mSpanned;
 
+  /**
+   * Used to collect some text size affecting attributes to fix some text cut-off issues when users
+   * adjust text size and font weight to the max value in system font settings.
+   */
+  private TextAttributes mTextAttributes;
```
Comment (NickGerleman) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java:74]: It seems a bit confusing, to have most of the attributes like font size, letter spacing, as member variables, then also having TextAttributes with those same members, and sometimes we use one but not the other.

Could we either:
1. Remove the member variables for where we can use TextAttributes instead?
2. Keep using the member variables, and copy any logic we might be wanting to use?

**Review by NickGerleman — COMMENTED** [review]
Could you help point out which change here leads to the new behavior?

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java (hunk 1 above)
Comment (jcdhlzq) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java]: > Why was this changed to `pt`? Input should be SP/dips.

It's just a bad name, I'll change it.

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java (hunk 2 above)
Comment (jcdhlzq) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java:72]: > It seems a bit confusing, to have most of the attributes like font size, letter spacing, as member variables, then also having TextAttributes with those same members, and sometimes we use one but not the other.
> 
> Could we either:
> 
> 1. Remove the member variables for where we can use TextAttributes instead?
> 2. Keep using the member variables, and copy any logic we might be wanting to use?

I'll remove the member variables that not needed now.

**Comment by jcdhlzq** [comment]
@ryancat One error happened on CircleCI, but I can not rerun it.

**Comment by jcdhlzq** [comment]
> Could you help point out which change here leads to the new behavior?

The key change is to set text size by calling `TextView#setTextSize()`. That will actually set text size to `mTextPaint` of TextView, which will be used as an input when measure&draw.

**Review by NickGerleman — APPROVED** [review]

**Comment by jcdhlzq** [comment]
> @ryancat One error happened on CircleCI, but I can not rerun it.

@ryancat This pr is reverted, what's the reason and what can I do to proceed the pr?

**Comment by NickGerleman** [comment]
It ended up causing a few hundred test failures due to NPEs. I will see if I can dig up the stack trace soon.

**Comment by NickGerleman** [comment]
```
09-26 12:49:01.153  4183  4183 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void com.facebook.react.views.text.TextAttributes.setLetterSpacing(float)' on a null object reference
09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at com.facebook.react.views.text.ReactTextView.setLetterSpacing(ReactTextView.java:668)
09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at android.widget.TextView.<init>(TextView.java:1411)
09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at android.widget.TextView.<init>(TextView.java:704)
09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at androidx.appcompat.widget.AppCompatTextView.<init>(AppCompatTextView.java:113)
09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at androidx.appcompat.widget.AppCompatTextView.<init>(AppCompatTextView.java:108)
09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at androidx.appcompat.widget.AppCompatTextView.<init>(AppCompatTextView.java:104)
09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at com.facebook.react.views.text.ReactTextView.<init>(ReactTextView.java:75)
09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at com.facebook.react.views.text.ReactTextViewManager.createViewInstance(ReactTextViewManager.java:82)
09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at com.facebook.react.views.text.ReactTextViewManager.createViewInstance(ReactTextViewManager.java:36)
09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at com.facebook.react.uimanager.ViewManager.createViewInstance(ViewManager.java:176)
```

**Comment by jcdhlzq** [comment]
> ```
> 09-26 12:49:01.153  4183  4183 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void com.facebook.react.views.text.TextAttributes.setLetterSpacing(float)' on a null object reference
> 09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at com.facebook.react.views.text.ReactTextView.setLetterSpacing(ReactTextView.java:668)
> 09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at android.widget.TextView.<init>(TextView.java:1411)
> 09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at android.widget.TextView.<init>(TextView.java:704)
> 09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at androidx.appcompat.widget.AppCompatTextView.<init>(AppCompatTextView.java:113)
> 09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at androidx.appcompat.widget.AppCompatTextView.<init>(AppCompatTextView.java:108)
> 09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at androidx.appcompat.widget.AppCompatTextView.<init>(AppCompatTextView.java:104)
> 09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at com.facebook.react.views.text.ReactTextView.<init>(ReactTextView.java:75)
> 09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at com.facebook.react.views.text.ReactTextViewManager.createViewInstance(ReactTextViewManager.java:82)
> 09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at com.facebook.react.views.text.ReactTextViewManager.createViewInstance(ReactTextViewManager.java:36)
> 09-26 12:49:01.153  4183  4183 E AndroidRuntime: 	at com.facebook.react.uimanager.ViewManager.createViewInstance(ViewManager.java:176)
> ```

ok, I will fix it.

**Comment by jcdhlzq** [comment]
It seems that this pr cannot reopen, so I make a new one #39711 @NickGerleman @ryancat

**Comment by jcdhlzq** [comment]
> It ended up causing a few hundred test failures due to NPEs. I will see if I can dig up the stack trace soon.

Hi, @ryancat 

The NPE is caused by overriding `setLetterSpacing` which is called in super constructor. I've named it with `setLetterSpacingAttr` in new PR #39711 .

And all checks have passed in #39711.  Shall we proceed with it?
